src/util.js

```diff
@@ -83,14 +83,20 @@ function serviceResponse(service, { serviceName = service.serviceName, action, a
   return client.executeServiceFunction(serviceName, action, args);
 }
 
+let platform = null;
+
 /**
  * Utility function to get the platform.
  *
  * @memberof util
  * @returns {Object} An object of the form {name: ..., version: ...}, or null
  */
 function getPlatform() {
-  return platform ? platform : null;
+  return platform ? platform : _platform;
+}
+
+function setPlatform(customPlatform) {
+  platform = customPlatform;
```

Comment:
relying on mutable global state in this module is very smelly, especially when it's the core SDK code reaching into a util module...
is it absolutely necessary to do this and/or is there a different approach?

Reply:
It is necessary to maintain a separate variable for platform, since `detect-browser` won't work on non-browsers (have we tested this on node?). That being said, it doesn't explicitly need to be tied to global state.

I made it global because platform shouldn't be mutable once it's set. `platform` should, in theory, be a static value. Can't really change the platform you're on. That being said, since Global Variables Are Bad™ (http://wiki.c2.com/?GlobalVariablesAreBad), there are other alternatives.

I think a creative alternative would be to lock the state once it's set via a tokenized dispatcher, e.g,:

```javascript
let isPlatformSet = false;
function setPlatform(customPlatform) {
    if (isPlatformSet) {
        throw Error("platform has already been set");
    }

    platform = customPlatform;
}
```

I suppose another option could be a singleton that must be initialized.

Reply:
With that said, a parameterized Singleton doesn't really work in practice, and would have to fallback on the some mechanisms.

Reply:
`detect-browser` works in node:
```
$ node -e 'console.log(require("detect-browser"))'
{ name: 'node', version: '8.6.0' }
```

I still don't understand why `platform` would ever need to be modified at all. If there's a need to derive some kind of `customPlatform` option that overrides what `getPlatform` normally returns, it should be dealt with in the caller. It looks like the use case is so that you can provide a `customPlatform` option to the client constructor. Since the platform is only actually ever used in `getDeviceInfo` which is in turn used by the auth providers, it seems like those auth methods should just take a `customPlatform` option that flows downward from the client class.

Reply:
I'm not sure detect-browser will work with react-native, as they supply their own module for platform: https://facebook.github.io/react-native/docs/platform-specific-code.html .